File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/javascript/SalazarProgrammer.js
// #01 OPERADORES Y ESTRUCTURAS DE CONTROL
// Dificultad: Fácil | Publicación: 02/01/24 | Corrección: 08/01/24

/*
 * EJERCICIO:
 * - Crea ejemplos utilizando todos los tipos de operadores de tu lenguaje:
 *   Aritméticos, lógicos, de comparación, asignación, identidad, pertenencia, bits...
 *   (Ten en cuenta que cada lenguaje puede poseer unos diferentes)
 * - Utilizando las operaciones con operadores que tú quieras, crea ejemplos
 *   que representen todos los tipos de estructuras de control que existan
 *   en tu lenguaje:
 *   Condicionales, iterativas, excepciones...
 * - Debes hacer print por consola del resultado de todos los ejemplos.
 *
 * DIFICULTAD EXTRA (opcional):
 * Crea un programa que imprima por consola todos los números comprendidos
 * entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
 *
 * Seguro que al revisar detenidamente las posibilidades has descubierto algo nuevo.
 */

/*
 * RESPUESTAS:
 * - Operadores y estructuras de control
 */

console.log('=== OPERADORES EN JAVASCRIPT ===');

// 1. OPERADORES ARITMÉTICOS
console.log('\n1. OPERADORES ARITMÉTICOS:');
let a = 15;
let b = 4;

console.log('a =', a, ', b =', b);
console.log('Suma (a + b):', a + b);
console.log('Resta (a - b):', a - b);
console.log('Multiplicación (a * b):', a * b);
console.log('División (a / b):', a / b);
console.log('Módulo (a % b):', a % b);
console.log('Exponenciación (a ** b):', a ** b);
console.log('Incremento (a++):', a++); // Post-incremento
console.log('Decremento (--b):', --b); // Pre-decremento

// 2. OPERADORES DE COMPARACIÓN
console.log('\n2. OPERADORES DE COMPARACIÓN:');
console.log('Igualdad (a == 15):', a == 15);
console.log("Igualdad estricta (a === '15'):", a === '15');
console.log('Desigualdad (a != 15):', a != 15);
console.log('Desigualdad estricta (a !== 15):', a !== 15);
console.log('Mayor que (a > 10):', a > 10);
console.log('Menor que (a < 20):', a < 20);
console.log('Mayor o igual (a >= 15):', a >= 15);
console.log('Menor o igual (a <= 15):', a <= 15);

// 3. OPERADORES LÓGICOS
console.log('\n3. OPERADORES LÓGICOS:');
let x = true;
let y = false;

console.log('AND (x && y):', x && y);
console.log('OR (x || y):', x || y);
console.log('NOT (!x):', !x);
console.log("Nullish Coalescing (null ?? 'default'):", null ?? 'default');
console.log('Optional Chaining (obj?.prop):', {}.prop?.nested);

// 4. OPERADORES DE ASIGNACIÓN
console.log('\n4. OPERADORES DE ASIGNACIÓN:');
let c = 10;
console.log('Asignación (=): c =', c);
c += 5;
console.log('Suma y asignación (+=):', c);
c -= 3;
console.log('Resta y asignación (-=):', c);
c *= 2;
console.log('Multiplicación y asignación (*=):', c);
c /= 4;
console.log('División y asignación (/=):', c);
c %= 3;
console.log('Módulo y asignación (%=):', c);

// 5. OPERADORES DE BITS
console.log('\n5. OPERADORES DE BITS:');
let num1 = 5; // 0101 en binario
let num2 = 3; // 0011 en binario

console.log('AND (&):', num1 & num2); // 0001 = 1
console.log('OR (|):', num1 | num2); // 0111 = 7
console.log('XOR (^):', num1 ^ num2); // 0110 = 6
console.log('NOT (~num1):', ~num1); // Complemento a 2
console.log('Desplazamiento izquierda (num1 << 1):', num1 << 1); // 1010 = 10
console.log('Desplazamiento derecha (num1 >> 1):', num1 >> 1); // 0010 = 2

// 6. OPERADORES DE IDENTIDAD/TIPO
console.log('\n6. OPERADORES DE IDENTIDAD/TIPO:');
console.log('typeof 42:', typeof 42);
console.log("typeof 'hello':", typeof 'hello');
console.log('instanceof ([] instanceof Array):', [] instanceof Array);
console.log("in ('length' in []):", 'length' in []);

// 7. OPERADORES DE PERTENENCIA (para arrays y objetos)
console.log('\n7. OPERADORES DE PERTENENCIA:');
const array = [1, 2, 3, 4, 5];
const objeto = { nombre: 'Juan', edad: 30 };

console.log("in operator ('edad' in objeto):", 'edad' in objeto);
console.log('Array includes (array.includes(3)):', array.includes(3));

console.log('\n=== ESTRUCTURAS DE CONTROL ===');

// 1. CONDICIONALES
console.log('\n1. ESTRUCTURAS CONDICIONALES:');

// if-else
if (a > 10) {
	console.log('if: a es mayor que 10');
} else {
	console.log('if: a no es mayor que 10');
}

// if-else if-else
if (a > 20) {
	console.log('a > 20');
} else if (a > 15) {
	console.log('15 < a <= 20');
} else {
	console.log('a <= 15');
}

// switch
let dia = 3;
switch (dia) {
	case 1:
		console.log('switch: Lunes');
		break;
	case 2:
		console.log('switch: Martes');
		break;
	case 3:
		console.log('switch: Miércoles');
		break;
	default:
		console.log('switch: Otro día');
}

// Operador ternario
let resultado = a > 10 ? 'Mayor que 10' : 'Menor o igual a 10';
console.log('Ternario:', resultado);

// 2. ESTRUCTURAS ITERATIVAS
console.log('\n2. ESTRUCTURAS ITERATIVAS:');

// for
console.log('For loop:');
for (let i = 0; i < 5; i++) {
	console.log('  Iteración:', i);
}

// while
console.log('While loop:');
let contador = 0;
while (contador < 3) {
	console.log('  Contador:', contador);
	contador++;
}

// do-while
console.log('Do-while loop:');
contador = 0;
do {
	console.log('  Contador:', contador);
	contador++;
} while (contador < 3);

// for...of (arrays)
console.log('For...of (arrays):');
for (const elemento of array) {
	console.log('  Elemento:', elemento);
}

// for...in (objetos)
console.log('For...in (objetos):');
for (const clave in objeto) {
	console.log(`  ${clave}: ${objeto[clave]}`);
}

// 3. MANEJO DE EXCEPCIONES
console.log('\n3. MANEJO DE EXCEPCIONES:');

try {
	console.log('Intentando operación...');
	let division = 10 / 0;
	if (!isFinite(division)) {
		throw new Error('División por cero o resultado infinito');
	}
	console.log('Resultado:', division);
} catch (error) {
	console.log('Error capturado:', error.message);
} finally {
	console.log('Bloque finally ejecutado');
}

// 4. CONTROL DE FLUJO
console.log('\n4. CONTROL DE FLUJO:');

// break
console.log('Break en loop:');
for (let i = 0; i < 10; i++) {
	if (i === 5) break;
	console.log('  i =', i);
}

// continue
console.log('Continue en loop:');
for (let i = 0; i < 5; i++) {
	if (i === 2) continue;
	console.log('  i =', i);
}

// return (en función)
function ejemploReturn() {
	console.log('Esta línea se ejecuta');
	return 'Valor retornado';
}
console.log('Return:', ejemploReturn());

/*
 * DIFICULTAD EXTRA (opcional):
 * Números entre 10 y 55, pares, que no son 16 ni múltiplos de 3
 */
console.log('\n=== DIFICULTAD EXTRA ===');

function numerosEspeciales() {
	let count = 0;

	for (let i = 10; i <= 55; i++) {
		if (i % 2 === 0 && i !== 16 && i % 3 !== 0) {
			console.log(i);
			count++;
		}
	}

	return count;
}

const total = numerosEspeciales();
console.log(`\nTotal de números encontrados: ${total}`);

File: Roadmap/01 - OPERADORES Y ESTRUCTURAS DE CONTROL/haskell/SalazarProgrammer.hs
{-
  EJERCICIO:
  - Operadores: Aritméticos, lógicos, de comparación, asignación, identidad, pertenencia, bits...
  - Estructuras de control: Condicionales, iterativas, excepciones...
  - Debes hacer print por consola del resultado de todos los ejemplos.
-}

module Main where

-- Operadores Aritméticos
operadoresAritmeticos :: IO ()
operadoresAritmeticos = do
    putStrLn "\n--- OPERADORES ARITMÉTICOS ---"
    putStrLn $ "5 + 3 = " ++ show (5 + 3)      -- Suma: 8
    putStrLn $ "5 - 3 = " ++ show (5 - 3)      -- Resta: 2
    putStrLn $ "5 * 3 = " ++ show (5 * 3)      -- Multiplicación: 15
    putStrLn $ "10 / 3 = " ++ show (10 / 3)    -- División: 3.333...
    putStrLn $ "10 `div` 3 = " ++ show (10 `div` 3)  -- División entera: 3
    putStrLn $ "10 `mod` 3 = " ++ show (10 `mod` 3)  -- Módulo: 1
    putStrLn $ "2 ^ 3 = " ++ show (2 ^ 3)      -- Potencia: 8

-- Operadores de Comparación
operadoresComparacion :: IO ()
operadoresComparacion = do
    putStrLn "\n--- OPERADORES DE COMPARACIÓN ---"
    putStrLn $ "5 == 3 = " ++ show (5 == 3)    -- Igualdad: False
    putStrLn $ "5 /= 3 = " ++ show (5 /= 3)    -- Desigualdad: True
    putStrLn $ "5 > 3 = " ++ show (5 > 3)      -- Mayor que: True
    putStrLn $ "5 < 3 = " ++ show (5 < 3)      -- Menor que: False
    putStrLn $ "5 >= 5 = " ++ show (5 >= 5)    -- Mayor o igual: True
    putStrLn $ "5 <= 3 = " ++ show (5 <= 3)    -- Menor o igual: False

-- Operadores Lógicos
operadoresLogicos :: IO ()
operadoresLogicos = do
    putStrLn "\n--- OPERADORES LÓGICOS ---"
    putStrLn $ "True && False = " ++ show (True && False)  -- AND: False
    putStrLn $ "True || False = " ++ show (True || False)  -- OR: True
    putStrLn $ "not True = " ++ show (not True)            -- NOT: False

-- Operadores a Nivel de Bits
operadoresBits :: IO ()
operadoresBits = do
    putStrLn "\n--- OPERADORES A NIVEL DE BITS ---"
    putStrLn $ "5 .&. 3 = " ++ show (5 .&. 3)  -- AND bit a bit: 1
    putStrLn $ "5 .|. 3 = " ++ show (5 .|. 3)  -- OR bit a bit: 7
    putStrLn $ "5 `xor` 3 = " ++ show (5 `xor` 3)  -- XOR: 6
    putStrLn $ "complement (5 :: Int) = " ++ show (complement (5 :: Int))  -- Complemento: -6

-- En Haskell no hay operadores de asignación como en lenguajes imperativos
-- Las variables son inmutables por defecto

-- Estructuras de Control Condicionales
estructurasCondicionales :: IO ()
estructurasCondicionales = do
    putStrLn "\n--- ESTRUCTURAS CONDICIONALES ---"

    -- If-Then-Else
    let resultado1 = if 5 > 3 then "Mayor" else "Menor"
    putStrLn $ "If-Then-Else: " ++ resultado1

    -- Guardas (Guards)
    let clasificar x
            | x > 10   = "Grande"
            | x > 5    = "Mediano"
            | otherwise = "Pequeño"
    putStrLn $ "Guardas (7): " ++ clasificar 7

    -- Pattern Matching
    let factorial 0 = 1
        factorial n = n * factorial (n - 1)
    putStrLn $ "Pattern Matching (factorial 5): " ++ show (factorial 5)

-- Estructuras Iterativas (Haskell usa recursión en lugar de bucles)
estructurasIterativas :: IO ()
estructurasIterativas = do
    putStrLn "\n--- ESTRUCTURAS ITERATIVAS ---"

    -- Recursión
    let sumaLista [] = 0
        sumaLista (x:xs) = x + sumaLista xs
    putStrLn $ "Recursión (suma [1,2,3,4]): " ++ show (sumaLista [1,2,3,4])

    -- Map (transformación)
    let duplicar = map (*2)
    putStrLn $ "Map (duplicar [1,2,3]): " ++ show (duplicar [1,2,3])

    -- Filter (filtrado)
    let pares = filter even
    putStrLn $ "Filter (pares [1,2,3,4,5]): " ++ show (pares [1,2,3,4,5])

    -- Fold (acumulación)
    let producto = foldl (*) 1
    putStrLn $ "Fold (producto [1,2,3,4]): " ++ show (producto [1,2,3,4])

-- Manejo de Excepciones
estructurasExcepciones :: IO ()
estructurasExcepciones = do
    putStrLn "\n--- MANEJO DE EXCEPCIONES ---"

    -- Usando Maybe
    let divisionSegura _ 0 = Nothing
        divisionSegura a b = Just (a / b)

    case divisionSegura 10 2 of
        Just resultado -> putStrLn $ "División segura (10/2): " ++ show resultado
        Nothing -> putStrLn "Error: división por cero"

    case divisionSegura 10 0 of
        Just resultado -> putStrLn $ "División segura: " ++ show resultado
        Nothing -> putStrLn "Error: división por cero (manejado con Maybe)"

{-
  DIFICULTAD EXTRA (opcional):
 * Crea un programa que imprima por consola todos los números comprendidos
 * entre 10 y 55 (incluidos), pares, y que no son ni el 16 ni múltiplos de 3.
-}

-- DIFICULTAD EXTRA
dificultadExtra :: IO ()
dificultadExtra = do
    putStrLn "\n--- DIFICULTAD EXTRA ---"
    putStrLn "Números entre 10 y 55 (pares, no 16, no múltiplos de 3):"

    let numeros = [x | x <- [10..55], even x, x /= 16, x `mod` 3 /= 0]
    mapM_ (putStrLn . show) numeros
    putStrLn $ "Total: " ++ show (length numeros) ++ " números"

-- Función principal
main :: IO ()
main = do
    operadoresAritmeticos
    operadoresComparacion
    operadoresLogicos
    operadoresBits
    estructurasCondicionales
    estructurasIterativas
    estructurasExcepciones
    dificultadExtra

-- FIN
